crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1107,6 +1123,104 @@ impl<'db> TypeInferenceBuilder<'db> {
+                            && if_stmt
+                                .elif_else_clauses
+                                .iter()
+                                .all(|clause| does_suite_always_return(&clause.body))
+                    }
+                    Some(ast::Stmt::While(while_stmt)) => {
+                        does_suite_always_return(&while_stmt.body)
+                            && does_suite_always_return(&while_stmt.orelse)
+                    }
+                    Some(ast::Stmt::For(for_stmt)) => {
+                        does_suite_always_return(&for_stmt.body)
+                            && does_suite_always_return(&for_stmt.orelse)
+                    }
+                    Some(ast::Stmt::Try(try_stmt)) => {
+                        does_suite_always_return(&try_stmt.body)
+                            && does_suite_always_return(&try_stmt.orelse)
+                            && does_suite_always_return(&try_stmt.finalbody)
+                            && try_stmt.handlers.iter().all(|handler| {
+                                does_suite_always_return(&handler.as_except_handler().unwrap().body)
+                            })
+                    }
+                    Some(ast::Stmt::With(with_stmt)) => does_suite_always_return(&with_stmt.body),
+                    Some(ast::Stmt::Match(match_stmt)) => match_stmt
+                        .cases
+                        .iter()
+                        .all(|case| does_suite_always_return(&case.body)),
+                    _ => false,
+                }
+            }
+            let is_overloaded = function.decorator_list.iter().any(|decorator| {
+                let decorator_type = self.file_expression_type(&decorator.expression);
+
+                decorator_type
+                    .into_function_literal()
+                    .is_some_and(|f| f.is_known(self.db(), KnownFunction::Overload))
+            });
+            // TODO: Protocol / abstract methods can have empty bodies
+            if (self.in_stub() || is_overloaded)
+                && self.return_types_and_ranges.is_empty()
+                && is_stub_suite(&function.body)
+            {
+                return;
+            }
+            for invalid in self
+                .return_types_and_ranges
+                .iter()
+                .filter(|ty_range| !ty_range.ty.is_assignable_to(self.db(), declared_ty))
+            {
+                report_invalid_return_type(
+                    &self.context,
+                    invalid.range,
+                    function.returns.as_ref().unwrap().range(),
+                    declared_ty,
+                    invalid.ty,
+                );
+            }
+            let scope_id = self.index.node_scope(NodeWithScopeRef::Function(function));
+            let use_def = self.index.use_def_map(scope_id);
+            if use_def.can_implicit_return(self.db())
+                && !does_suite_always_return(&function.body)
```

Comment:
We shouldn't need this function at all; this is what `can_implicit_return` should do for us. I investigated why it doesn't, and realized that it's our temporary hack here: https://github.com/astral-sh/ruff/blob/main/crates/red_knot_python_semantic/src/semantic_index/builder.rs#L880-L899

Really sorry I missed this and it caused a bunch of extra unnecessary work for you! I found a 2-3 line temporary fix that will solve this until we get rid of this hack, will push that up to this PR since I already did it.